File: src/aws/ec2.rs
use std::time::Duration;

use anyhow::{Result, anyhow};
use aws_sdk_ec2::types::Filter;
use crossbeam::utils::Backoff;
use log::debug;
use tokio::runtime::Handle;

use crate::vmspec::EbsVolumeAttachment;

#[derive(Debug)]
pub struct Ec2Client {
    rt: Handle,
    client: Ec2ClientAsync,
}

impl Ec2Client {
    pub fn new(rt: Handle, client: aws_sdk_ec2::Client) -> Self {
        let client_async = Ec2ClientAsync::new(client);
        Self {
            rt,
            client: client_async,
        }
    }

    pub fn ensure_ebs_volume_attached(
        &self,
        attachment: &EbsVolumeAttachment,
        device: &str,
        availability_zone: &str,
        instance_id: &str,
    ) -> Result<()> {
        self.rt.block_on(self.client.ensure_ebs_volume_attached(
            attachment,
            device,
            availability_zone,
            instance_id,
        ))
    }
}

#[derive(Debug)]
pub struct Ec2ClientAsync {
    client: aws_sdk_ec2::Client,
}

impl Ec2ClientAsync {
    pub fn new(client: aws_sdk_ec2::Client) -> Self {
        Self { client }
    }

    pub async fn ensure_ebs_volume_attached(
        &self,
        attachment: &EbsVolumeAttachment,
        device: &str,
        availability_zone: &str,
        instance_id: &str,
    ) -> Result<()> {
        if self
            .ebs_volume_attached(attachment, device, instance_id)
            .await?
        {
            return Ok(());
        }
        let volume_id = self
            .wait_for_ebs_volume(attachment, availability_zone)
            .await?;
        self.client
            .attach_volume()
            .device(device)
            .instance_id(instance_id)
            .volume_id(volume_id)
            .send()
            .await
            .map_err(|e| anyhow!("unable to attach EBS volume: {}", e.into_service_error()))?;
        Ok(())
    }

    async fn ebs_volume_attached(
        &self,
        attachment: &EbsVolumeAttachment,
        device: &str,
        instance_id: &str,
    ) -> Result<bool> {
        let mut desc_vol = self
            .client
            .describe_volumes()
            .filters(
                Filter::builder()
                    .name("attachment.instance-id")
                    .values(instance_id)
                    .build(),
            )
            .filters(
                Filter::builder()
                    .name("attachment.device")
                    .values(device)
                    .build(),
            );
        for tag in &attachment.tags {
            let filters = if let Some(ref value) = tag.value {
                Filter::builder()
                    .name(format!("tag:{}", tag.key))
                    .values(value)
                    .build()
            } else {
                Filter::builder()
                    .name("tag-key")
                    .values(tag.key.clone())
                    .build()
            };
            desc_vol = desc_vol.filters(filters);
        }
        match desc_vol.clone().send().await {
            Ok(vol_out) => {
                if let Some(ref vols) = vol_out.volumes {
                    return Ok(!vols.is_empty());
                }
            }
            Err(e) => {
                return Err(anyhow!(
                    "error describing EBS volumes: {}",
                    e.into_service_error()
                ));
            }
        }
        Ok(false)
    }

    async fn wait_for_ebs_volume(
        &self,
        attachment: &EbsVolumeAttachment,
        availability_zone: &str,
    ) -> Result<String> {
        let mut desc_vol = self
            .client
            .describe_volumes()
            .filters(Filter::builder().name("status").values("available").build())
            .filters(
                Filter::builder()
                    .name("availability-zone")
                    .values(availability_zone)
                    .build(),
            );
        for tag in &attachment.tags {
            let filters = if let Some(ref value) = tag.value {
                Filter::builder()
                    .name(format!("tag:{}", tag.key))
                    .values(value)
                    .build()
            } else {
                Filter::builder()
                    .name("tag-key")
                    .values(tag.key.clone())
                    .build()
            };
            desc_vol = desc_vol.filters(filters);
        }
        let start = std::time::Instant::now();
        let timeout = Duration::from_secs(attachment.timeout.unwrap_or(300));
        let backoff = Backoff::new();
        loop {
            let result = desc_vol.clone().send().await;
            match result {
                Err(e) => debug!("error describing EBS volumes: {}", e.into_service_error()),
                Ok(vol_out) => {
                    if let Some(ref vols) = vol_out.volumes
                        && let Some(volume) = vols.first()
                            && let Some(volume_id) = &volume.volume_id {
                                debug!("found matching EBS volume: {:?}", volume);
                                return Ok(volume_id.clone());
                            }
                    debug!("no EBS volume found matching filters");
                }
            }
            if start.elapsed() > timeout {
                return Err(anyhow!("timeout waiting for EBS volume to be available"));
            }
            debug!("waiting for EBS volume to be available");
            backoff.snooze();
        }
    }
}

File: src/aws/ec2.zig
//! EC2 client for EBS volume operations.

const std = @import("std");
const Allocator = std.mem.Allocator;
const testing = std.testing;

const aws = @import("aws");
const ec2 = @import("ec2");
const Filter = ec2.types.Filter;

const backoff = @import("../backoff.zig");
const vmspec = @import("../vmspec.zig");
const EbsVolumeAttachment = vmspec.EbsVolumeAttachment;
const AwsTag = vmspec.AwsTag;

const scoped_log = std.log.scoped(.aws_ec2);

pub const Ec2Error = error{
    RequestFailed,
    ServiceError,
    Timeout,
};

/// Helper for building EC2 filter lists with automatic memory management.
const FilterBuilder = struct {
    allocator: Allocator,
    filters: std.ArrayListUnmanaged(Filter) = .empty,
    values: std.ArrayListUnmanaged([]const []const u8) = .empty,
    names: std.ArrayListUnmanaged([]const u8) = .empty,

    fn init(allocator: Allocator) FilterBuilder {
        return .{ .allocator = allocator };
    }

    fn deinit(self: *FilterBuilder) void {
        for (self.names.items) |n| self.allocator.free(n);
        self.names.deinit(self.allocator);
        for (self.values.items) |arr| self.allocator.free(arr);
        self.values.deinit(self.allocator);
        self.filters.deinit(self.allocator);
    }

    fn add(self: *FilterBuilder, name: []const u8, value: []const u8) !void {
        const vals = try self.allocator.dupe([]const u8, &[_][]const u8{value});
        try self.values.append(self.allocator, vals);
        try self.filters.append(self.allocator, .{ .name = name, .values = vals });
    }

    /// Like add, but takes ownership of an allocated name.
    fn addAlloc(self: *FilterBuilder, name: []const u8, value: []const u8) !void {
        try self.names.append(self.allocator, name);
        try self.add(name, value);
    }

    fn items(self: *FilterBuilder) []Filter {
        return self.filters.items;
    }
};

pub const Ec2Client = struct {
    allocator: Allocator,
    config: aws.Config,

    const Self = @This();

    pub fn init(allocator: Allocator, region: []const u8) !Self {
        return Self{
            .allocator = allocator,
            .config = try aws.Config.load(allocator, .{ .region = region }),
        };
    }

    pub fn deinit(self: *Self) void {
        self.config.deinit();
    }

    /// Ensure an EBS volume is attached to the instance.
    /// If the volume is already attached, returns immediately.
    /// If not, finds an available volume matching the tags and attaches it.
    pub fn ensureVolumeAttached(
        self: *Self,
        attachment: *const EbsVolumeAttachment,
        device: []const u8,
        availability_zone: []const u8,
        instance_id: []const u8,
    ) !void {
        if (try self.isVolumeAttached(attachment, device, instance_id)) {
            scoped_log.debug("volume already attached at {s}", .{device});
            return;
        }

        const volume_id = try self.waitForAvailableVolume(attachment, availability_zone);
        defer self.allocator.free(volume_id);

        scoped_log.info("attaching volume {s} to {s}", .{ volume_id, device });

        try self.attachVolume(volume_id, device, instance_id);
    }

    /// Check if a volume matching the attachment spec is already attached.
    fn isVolumeAttached(
        self: *Self,
        attachment: *const EbsVolumeAttachment,
        device: []const u8,
        instance_id: []const u8,
    ) !bool {
        var client = ec2.Client.init(self.allocator, &self.config);
        defer client.deinit();

        var arena = std.heap.ArenaAllocator.init(self.allocator);
        defer arena.deinit();

        var fb = FilterBuilder.init(self.allocator);
        defer fb.deinit();

        try fb.add("attachment.instance-id", instance_id);
        try fb.add("attachment.device", device);
        try self.addTagFilters(&fb, attachment.tags);

        var diagnostic: ec2.ServiceError = undefined;

        const result = client.describeVolumes(
            arena.allocator(),
            .{ .filters = fb.items() },
            .{ .diagnostic = &diagnostic },
        ) catch |err| {
            if (err == error.ServiceError) {
                defer diagnostic.deinit();
                scoped_log.err(
                    "DescribeVolumes failed: {s}: {s}",
                    .{ diagnostic.code(), diagnostic.message() },
                );
                return Ec2Error.ServiceError;
            }
            scoped_log.err("DescribeVolumes failed: {s}", .{@errorName(err)});
            return Ec2Error.RequestFailed;
        };

        if (result.volumes) |volumes| {
            return volumes.len > 0;
        }

        return false;
    }

    /// Wait for an available volume matching the attachment spec.
    /// Returns the volume ID when found.
    fn waitForAvailableVolume(
        self: *Self,
        attachment: *const EbsVolumeAttachment,
        availability_zone: []const u8,
    ) ![]const u8 {
        const timeout_secs = attachment.timeout orelse 300;
        const timeout_ns: u64 = timeout_secs * std.time.ns_per_s;
        const start_time = std.time.nanoTimestamp();

        var retry = backoff.RetryBackoff.init(10000);

        while (true) {
            if (try self.findAvailableVolume(attachment, availability_zone)) |volume_id| {
                return volume_id;
            }

            const elapsed: u64 = @intCast(std.time.nanoTimestamp() - start_time);
            if (elapsed > timeout_ns) {
                scoped_log.err("timeout waiting for EBS volume to be available", .{});
                return Ec2Error.Timeout;
            }

            scoped_log.debug("waiting for EBS volume to be available", .{});
            retry.wait();
        }
    }

    /// Find an available volume matching the attachment spec.
    fn findAvailableVolume(
        self: *Self,
        attachment: *const EbsVolumeAttachment,
        availability_zone: []const u8,
    ) !?[]const u8 {
        var client = ec2.Client.init(self.allocator, &self.config);
        defer client.deinit();

        var arena = std.heap.ArenaAllocator.init(self.allocator);
        defer arena.deinit();

        var fb = FilterBuilder.init(self.allocator);
        defer fb.deinit();

        try fb.add("status", "available");
        try fb.add("availability-zone", availability_zone);
        try self.addTagFilters(&fb, attachment.tags);

        var diagnostic: ec2.ServiceError = undefined;

        const result = client.describeVolumes(
            arena.allocator(),
            .{ .filters = fb.items() },
            .{ .diagnostic = &diagnostic },
        ) catch |err| {
            if (err == error.ServiceError) {
                defer diagnostic.deinit();
                scoped_log.err("DescribeVolumes failed: {s}: {s}", .{ diagnostic.code(), diagnostic.message() });
                return Ec2Error.ServiceError;
            }
            scoped_log.debug("DescribeVolumes failed: {s}", .{@errorName(err)});
            return null;
        };

        if (result.volumes) |volumes| {
            if (volumes.len > 0) {
                if (volumes[0].volume_id) |vol_id| {
                    scoped_log.debug("found matching EBS volume: {s}", .{vol_id});
                    return try self.allocator.dupe(u8, vol_id);
                }
            }
        }

        return null;
    }

    /// Attach a volume to an instance.
    fn attachVolume(
        self: *Self,
        volume_id: []const u8,
        device: []const u8,
        instance_id: []const u8,
    ) !void {
        var client = ec2.Client.init(self.allocator, &self.config);
        defer client.deinit();

        var arena = std.heap.ArenaAllocator.init(self.allocator);
        defer arena.deinit();

        var diagnostic: ec2.ServiceError = undefined;

        _ = client.attachVolume(
            arena.allocator(),
            .{
                .device = device,
                .instance_id = instance_id,
                .volume_id = volume_id,
            },
            .{ .diagnostic = &diagnostic },
        ) catch |err| {
            if (err == error.ServiceError) {
                defer diagnostic.deinit();
                scoped_log.err("AttachVolume failed: {s}: {s}", .{ diagnostic.code(), diagnostic.message() });
                return Ec2Error.ServiceError;
            }
            scoped_log.err("AttachVolume failed: {s}", .{@errorName(err)});
            return Ec2Error.RequestFailed;
        };

        scoped_log.info(
            "attached volume {s} to instance {s} at {s}",
            .{ volume_id, instance_id, device },
        );
    }

    fn addTagFilters(self: *Self, fb: *FilterBuilder, tags: []const AwsTag) !void {
        for (tags) |tag| {
            if (tag.value) |value| {
                const name = try std.fmt.allocPrint(self.allocator, "tag:{s}", .{tag.key});
                try fb.addAlloc(name, value);
            } else {
                try fb.add("tag-key", tag.key);
            }
        }
    }
};

test "FilterBuilder empty" {
    var fb = FilterBuilder.init(testing.allocator);
    defer fb.deinit();

    try testing.expectEqual(@as(usize, 0), fb.items().len);
}

test "FilterBuilder single filter" {
    var fb = FilterBuilder.init(testing.allocator);
    defer fb.deinit();

    try fb.add("status", "available");

    const filters = fb.items();
    try testing.expectEqual(@as(usize, 1), filters.len);
    try testing.expectEqualStrings("status", filters[0].name.?);
    try testing.expectEqual(@as(usize, 1), filters[0].values.?.len);
    try testing.expectEqualStrings("available", filters[0].values.?[0]);
}

test "FilterBuilder multiple filters" {
    var fb = FilterBuilder.init(testing.allocator);
    defer fb.deinit();

    try fb.add("status", "available");
    try fb.add("availability-zone", "us-east-1a");
    try fb.add("tag:Environment", "production");

    const filters = fb.items();
    try testing.expectEqual(@as(usize, 3), filters.len);

    try testing.expectEqualStrings("status", filters[0].name.?);
    try testing.expectEqualStrings("available", filters[0].values.?[0]);

    try testing.expectEqualStrings("availability-zone", filters[1].name.?);
    try testing.expectEqualStrings("us-east-1a", filters[1].values.?[0]);

    try testing.expectEqualStrings("tag:Environment", filters[2].name.?);
    try testing.expectEqualStrings("production", filters[2].values.?[0]);
}
